Deploy using private registry (#847)
knative/docs: serving/deploying-with-private-registry.md

# Deploying to Knative using a private container registry

> This guide walks you through deploying an application to Knative from source code in a git repository using a private container registry for the container image. The source code should contain a dockerfile. For this guide, we'll use this [simple app](https://github.com/mchmarny/simple-app), but you could use your own.

Reviewer: I think we should build one of the sample apps from within the Knative project. For example, this one uses one the Hello World sample in Go: https://github.com/knative/docs/blob/master/install/getting-started-knative-app.md

Author: Agree! We should considering doing this for some of the other samples that use external apps as well.

> ## Set up a private container registry and obtain credentials.

Reviewer: ```suggestion
## Set up a private container registry and obtain credentials
```

If you do not want your container image to be publicly available, you may want to use a private container registry. In this example, we'll use IBM Container Registry, but most of these concepts will be similar for other clouds.

1. Ensure you have the [IBM Cloud CLI](https://cloud.ibm.com/docs/cli/reference/ibmcloud/download_cli.html#install_use) installed.  

1. Install the container registry plugin:

    ```
    ibmcloud plugin install container-registry
    ```

1. Choose a name for your first namespace, and then create it. A namespace represents the spot within a registry that holds your images. You can set up multiple namespaces as well as control access to your namespaces by using IAM policies.

    ```

>     ibmcloud cr namespace-add <my_namespace>

Reviewer: Move command up so that it immediately follows the instruction, "Choose a name for your first namespace, and then create it." 

I realized this happens in a few places; I'll mark them, but not explain as thoroughly as I did here and the comment further down.

Reviewer: Use a semicolon to link your instructional sentence to the command. (Also applies more broadly.)

Author: Thanks, updated.

    ```

1. Create a token. The automated build processes you'll be setting up will use this token to access your images.

    ```

>     ibmcloud cr token-add --description "token description" --non-expiring --readwrite

Reviewer: Move command closer to instruction; put background info after.

    ```

> 1. The CLI output should include a token identifier and the token. Make note of the token. You can verify that the token was created by listing all tokens.

Reviewer: ```suggestion
1. The CLI output should include a token identifier and the token. Make note of the token. You can verify that the token was created by listing all tokens:
```

    ```

1. Update the "password" with your token_value>. Note that username will be the string `token`. Save the file.

1. Apply the secret to your cluster.

    ```
    kubectl apply --filename registry-push-secret.yaml
    ```

1. You will also need a secret for the knative-serving component to pull down an image from the private container registry. This secret will be a `docker-registry` type secret. You can create this via the commandline. For username, simply use the string `token`. For <token_value>, use the token you made note of earlier.

    ```
    kubectl create secret docker-registry ibm-cr-secret --docker-server=https://registry.ng.bluemix.net --docker-username=token --docker-password=<token_value>
    ```

A Service Account provides an identity for processes that run in a Pod. This Service Account will be used to link the build process for Knative to the Secrets you just created.

1. Create a file named service-account.yaml containing the following .yaml.

    ```
    apiVersion: v1
    kind: ServiceAccount
    metadata:
      name: build-bot
    secrets:
    - name: registry-push-secret
    imagePullSecrets:
    - name: ibm-cr-secret
    ```

> ## Deploy to Knative using the credentials

Reviewer: ```suggestion
## Deploy to Knative
```

> To build our application from the source on github, and push the resulting image to the IBM Container Registry, we will use the Kaniko build template.

Reviewer: ```suggestion
To build our application from the source on GitHub, and push the resulting image to the IBM Container Registry, we will use the Kaniko build template.
```

    ```
    apiVersion: serving.knative.dev/v1alpha1
    kind: Service
    metadata:
      name: app-from-source
      namespace: default
    spec:
      runLatest:
        configuration:
          build:
            apiVersion: build.knative.dev/v1alpha1
            kind: Build
            spec:
              serviceAccountName: build-bot
              source:
                git:
                  url: https://github.com/mchmarny/simple-app.git
                  revision: master
              template:
                name: kaniko
                arguments:
                - name: IMAGE
                  value: registry.ng.bluemix.net/{NAMESPACE}/app-from-source:latest
          revisionTemplate:
            spec:
              container:
                image: registry.ng.bluemix.net/{NAMESPACE}/app-from-source:latest 
                imagePullPolicy: Always
                env:
                - name: SIMPLE_MSG
                  value: "Hello from the sample app!"
    ```

> 1. Apply the configuration using `kubectl`. Applying this service definition will enable a number of events to happen:

Reviewer: ```suggestion
1. Apply the configuration using `kubectl`. Applying this service definition will kick off a series of events:
```

> - Fetch the revision specified from GitHub and build it into a container, using the Kaniko build template.

Reviewer: This unordered list needs to be further indented to avoid breaking the autonumbering on the ordered list. Right now, the numbered list items after this start over a 1.

> - Fetch the revision specified from GitHub and build it into a container, using the Kaniko build template.

Reviewer: Suggestion: rewrite the points in the unordered list to start with verbs, making it parallel.

Ex, "Fetches the revision...", "Pushes the latest image...", "Pulls down the latest image", "Starts the service..."

> - the latest image will be pushed to the private registry using the registry-push-secret

Reviewer: Capitalize the beginning of each list item.

- the latest image will be pulled down from the private registry using the ibm-cr-secret.
- the service will start, and your app will be running.

    ```

>     kubectl apply -f service.yaml

Reviewer: I suggest moving this code block up right after the sentence, "Apply the configuration using `kubectl`." Then you can have the explanation of what's happening after, without separating the instruction for the user and the command they're supposed to run.

    ```

1. You can run `kubectl get pods --watch` to see the pods initializing.

1. Once all the pods are initialized, you can see that your container image was built and pushed to the IBM Container Registry:

    ```
    ibmcloud cr image-list

>     ```

Reviewer: I'm wondering if it might make sense for this guide to also include the info for interacting with the app from the command line? 

Similar to: https://github.com/knative/docs/blob/master/install/getting-started-knative-app.md

Author: I was torn --  I felt that this is a more advanced tutorial, and it would just be repeating that information in 2 places. That said, it would be nice for them to confirm everything is working by testing the app.  I'll add to this PR to include that.

> # Deploying to Knative using a private container registry

Reviewer 2: How about update the title and file name as `Deploying to Knative using a private container registry on IBM Cloud`?

Author: I was hoping that the instructions were general purpose enough to be useful for other providers. The part I originally tripped up that motivated me to create this doc was creating the correct secrets. @samodell thoughts?

Reviewer: I'm happy with the current name. Up to you, @beemarie !

Author: I'll plan to keep it as is, and if we add another provider we can change the name to differentiate.

    ibmcloud cr namespace-add <my_namespace>
    ```
    
    A namespace represents the spot within a registry that holds your images. You can set up multiple namespaces as well as control access to your namespaces by using IAM policies.

1. Create a token:

    ```
    ibmcloud cr token-add --description "token description" --non-expiring --readwrite
    ```
    
    The automated build processes you'll be setting up will use this token to access your images.

1. The CLI output should include a token identifier and the token. Make note of the token. You can verify that the token was created by listing all tokens:

    ```
    ibmcloud cr token-list
    ```

## Provide container registry credentials to Knative
You will use the credentials you obtained in the previous section to authenticate to your private container registry. First, you'll need to create a secret to store the credentials for this registry. This secret will be used to push the built image to the container registry.

A Secret is a Kubernetes object containing sensitive data such as a password, a token, or a key. You can also read more about [Secrets](https://kubernetes.io/docs/concepts/configuration/secret/).

> 1. Create a file named registry-push-secret.yaml containing the following .yaml.

Reviewer 2: s/registry-push-secret.yaml/`registry-push-secret.yaml`

Reviewer 2: s/.yaml/content

Reviewer: ```suggestion
1. Create a file named `registry-push-secret.yaml` containing the following:
```

>     ```

Reviewer 2: Please add `yaml` to the end so as to highlight the syntax for YAML files, same for other YAML content in this file.

Example like this:

```yaml
    apiVersion: v1
    kind: Secret
    metadata:
      name: registry-push-secret
      annotations:
        build.knative.dev/docker-0: https://registry.ng.bluemix.net
    type: kubernetes.io/basic-auth
    stringData:
      username: token
      password: <token_value>
```

Reviewer: Created a suggested change to do this.

>     ```

Reviewer: ```suggestion
    ```yaml
```

    apiVersion: v1
    kind: Secret
    metadata:
      name: registry-push-secret
      annotations:
        build.knative.dev/docker-0: https://registry.ng.bluemix.net
    type: kubernetes.io/basic-auth
    stringData:
      username: token
      password: <token_value>
    ```

1. Update the "password" with your <token_value>. Note that username will be the string `token`. Save the file.

1. Apply the secret to your cluster.

    ```
    kubectl apply --filename registry-push-secret.yaml
    ```

1. You will also need a secret for the knative-serving component to pull down an image from the private container registry. This secret will be a `docker-registry` type secret. You can create this via the commandline. For username, simply use the string `token`. For <token_value>, use the token you made note of earlier.

    ```
    kubectl create secret docker-registry ibm-cr-secret --docker-server=https://registry.ng.bluemix.net --docker-username=token --docker-password=<token_value>
    ```

A Service Account provides an identity for processes that run in a Pod. This Service Account will be used to link the build process for Knative to the Secrets you just created.

> 1. Create a file named service-account.yaml containing the following .yaml.

Reviewer: ```suggestion
1. Create a file named `service-account.yaml` containing the following:
```

>     ```

Reviewer: ```suggestion
    ```yaml
```

    apiVersion: v1
    kind: ServiceAccount
    metadata:
      name: build-bot
    secrets:
    - name: registry-push-secret
    imagePullSecrets:
    - name: ibm-cr-secret
    ```

1. Apply the service account to your cluster:

    ```
    kubectl apply -f service-account.yaml
    ```

## Deploy to Knative
To build our application from the source on GitHub, and push the resulting image to the IBM Container Registry, we will use the Kaniko build template.

1. Install the Kaniko build template

    ```
    kubectl apply -f https://raw.githubusercontent.com/knative/build-templates/master/kaniko/kaniko.yaml
    ```

1. You need to create a service manifest which defines the service to deploy, including where the source code is and which build-template to use. Create a file named `service.yaml` and copy the following definition. Make sure to replace {NAMESPACE} with your own namespace you created earlier:

>     ```

Reviewer: ```suggestion
    ```yaml
```
